src/main/java/fi/aalto/cs/apluscourses/presentation/exercise/SubmissionViewModel.java

```diff
@@ -62,53 +63,53 @@ public String getPresentableExerciseName() {
   }
 
   @NotNull
-  public Group getSelectedGroup() {
-    return selectedGroup;
+  public SubmittableFile[] getFiles() {
+    return submissionInfo.getFiles();
   }
 
-  public void setGroup(@NotNull Group group) {
-    this.selectedGroup = group;
+  public int getCurrentSubmissionNumber() {
+    return submissionHistory.getNumberOfSubmissions() + 1;
   }
 
-  @NotNull
-  public List<String> getFilenames() {
-    return exercise.getFilenames();
+  public int getMaxNumberOfSubmissions() {
+    return submissionInfo.getSubmissionsLimit();
   }
 
-  public int getNumberOfSubmissions() {
-    return submissionHistory.getNumberOfSubmissions();
+  public Submission buildSubmission() {
+    Group group = Objects.requireNonNull(selectedGroup.get());
+    return new Submission(exercise, submissionInfo, files, group);
   }
 
-  public int getMaxNumberOfSubmissions() {
-    return exercise.getSubmissionsLimit();
+  public ValidationError validateSubmissionCount() {
+    return getCurrentSubmissionNumber() > getMaxNumberOfSubmissions()
+        ? new MaxNumberOfSubmissionsExceededError() : null;
```

Comment:
Does this prevent submissions or is it just a warning? If it prevents submission then we need to be sure that our max submissions takes any deviations into account

Reply:
I change it to a warning. I was already going to but thanks for reminding! :+1: